design/outlier-detection-design.md

```diff
@@ -0,0 +1,262 @@
+
+Below is the generated envoy configuration:
+```yaml
+cluster:
+  common_lb_config:
+    healthy_panic_threshold:
+      value: 0.0
+  outlier_detection:
+    enforcing_success_rate: 0
+    enforcing_consecutive_5xx: 0
+    max_ejection_percent: 100
+    split_external_local_origin_errors: true
+    consecutive_local_origin_failure: 5
+    #Default to 100% if not specified
+    #enforcing_consecutive_local_origin_failure: 100
+    enforcing_local_origin_success_rate: 0
+    enforcing_consecutive_gateway_failure: 0
+```
+
+### Example 4
+ ```yaml
+apiVersion: projectcontour.io/v1alpha1
+kind: HTTPProxy
+metadata:
+  name: local-external-origin-errors-and-consecutive-5xx
+  namespace: projectcontour
+spec:
+    virtualhost:
+        fqdn: outlierDetection.projectcontour.io
+    routes:
+      - conditions:
+          - prefix: /
+        services:
+          - name: s1
+            port: 80
+            outlierDetection:
+              consecutive5xxErrors: 10
+              maxEjectionPercent: 100
+              splitExternalLocalOriginErrors: true
+              consecutiveLocalOriginFailure: 5
+```
+In this example, when the s1 service continuously encounters 5 local origin errors or 10 external errors, the s1 service will be ejected. The maximum eviction rate is 100%. Panic is disabled, which is also the default value.
+
+Below is the generated envoy configuration:
+```yaml
+cluster:
+  common_lb_config:
+    healthy_panic_threshold:
+      value: 0.0
+  outlier_detection:
+    enforcing_success_rate: 0
+    consecutive_5xx: 10
+    enforcing_consecutive_5xx: 100
+    max_ejection_percent: 100
+    split_external_local_origin_errors: true
+    consecutive_local_origin_failure: 5
+    #Default to 100% if not specified
+    #enforcing_consecutive_local_origin_failure: 100
+    enforcing_local_origin_success_rate: 0
+    enforcing_consecutive_gateway_failure: 0
+``` 
+
+Notes:
+- If consecutive5xxErrors is specified as 0, and splitExternalLocalOriginErrors is true, then local errors will be ignored.This is especially useful when the upstream service explicitly returns a 5xx for some requests and you want to ignore those responses from upstream service while determining the outlier detection status of a host.
+- When accessing the upstream host through an opaque TCP connection, connection timeouts, connection errors, and request failures are all considered 5xx errors, and therefore these events are included in the 5xx error statistics.
+- Please refer to the [official documentation of Envoy][1] for more instructions.
+
+
+## Open Issues
+- Whether the existing supported configuration options meet the needs of the user.
+- Should a global switch be provided to record [ejection event logs][2]?
```

Comment:
I'm in favor of adding some discussion and exploration on this, I think with our active healthchecking configuration, it isn't ideal as-is but (IMO) people find it generally easier to explain and understand why their upstream is unhealthy (it returned an unhealthy status code from their healthcheck endpoint)

with passive healthchecking I can foresee some support requests etc. where users are confused as to why their services are deemed unhealthy, so having logs etc. to explain things would be great

we could also take advantage of the active healthchecking event log configuration as well

I think a great outcome of implementing this feature could be a debugging guide, including some mention of what stats to monitor (e.g. from https://www.envoyproxy.io/docs/envoy/latest/configuration/upstream/cluster_manager/cluster_stats#config-cluster-manager-cluster-stats-outlier-detection)

we would have to likely add a bootstrap flag to change the bootstrap `ClusterManager` configuration to enable logging

Reply:
Yes, adding a bootstrap flag to handle whether to enable logging can better help people confirm the cause of the failure. If it is considered feasible, I will create a new PR to implement this function.

Reply:
can also add an active health check flag to r[ecord health check failure events](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/health_check.proto#config-core-v3-healthcheck)